service/header/core_exchange.go

```diff
@@ -0,0 +1,122 @@
+package header
+
+import (
+	"context"
+	"math"
+
+	format "github.com/ipfs/go-ipld-format"
+
+	"github.com/tendermint/tendermint/pkg/da"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+	"github.com/tendermint/tendermint/types"
+
+	"github.com/celestiaorg/celestia-node/core"
+	"github.com/celestiaorg/celestia-node/ipld"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+type CoreExchange struct {
+	fetcher    *core.BlockFetcher
+	shareStore format.DAGService
+}
+
+func NewCoreExchange(fetcher *core.BlockFetcher, dag format.DAGService) *CoreExchange {
+	return &CoreExchange{
+		fetcher:    fetcher,
+		shareStore: dag,
+	}
+}
+
+func (ce *CoreExchange) RequestHeader(ctx context.Context, height uint64) (*ExtendedHeader, error) {
+	intHeight := int64(height) + 1
+	block, err := ce.fetcher.GetBlock(ctx, &intHeight)
+	if err != nil {
+		return nil, err
+	}
+	return ce.generateExtendedHeaderFromBlock(block)
+}
+
+func (ce *CoreExchange) RequestHeaders(ctx context.Context, origin, amount uint64) ([]*ExtendedHeader, error) {
```

Comment:
This will be so slooow, but it is fine for now. Just for reference, we should use BlockSearch API instead